# 선형 회귀 식 직접 만들기 

### 공부시간에 따른 시험 성적 

In [9]:
import numpy as np 

x=[2,4,6,8] #공부시간 
y=[81,93,91,97] #성적 

In [10]:
#x와 y의 평균값
mx=np.mean(x)
my=np.mean(y)
print("x의 평균값:", mx)
print("y의 평균값:", my)

x의 평균값: 5.0
y의 평균값: 90.5


![nn](Linear_Regression.png)

### 기울기

In [11]:
# 기울기 공식의 분모 
divisor=sum([(mx-i)**2 for i in x])

#기울기 공식의 분자 
def top(x,mx,y,my):
    d=0
    for i in range(len(x)):
        d+=(x[i]-mx)*(y[i]-my)
    return d 

dividend=top(x,mx,y,my)
print("분모:", divisor)
print("분자:",dividend)

분모: 20.0
분자: 46.0


### x기울기와 y절편 구하기 

In [12]:
a=dividend/divisor 
b=my-(mx*a)

print("기울기 a=",a)
print("y절편 b=",b)

기울기 a= 2.3
y절편 b= 79.0


# MSE, RMSE 평균 제곱근 오차 
### 오차를 평가하는 방법 
![nn](MSE.jpg)

In [13]:
# 기울기 a와 y절편을 3,76 이라하자 

ab=[3,76]

#x와 y 데이터 값 
data=[[2,81],[4,93],[6,91],[8,97]]
x=[i[0] for i in data]
y=[i[1] for i in data]

In [15]:
#y=ax+b에 a와 b값을 대입하여 결과를 출력하는 함수 
def predict(x):
    return ab[0]*x+ab[1]

#RMSE함수 
def rmse(p,a):
         return np.sqrt(((p-a)**2).mean())

#RMSE 함수를 각 y 값에 대입하여 최종 값을 구하는 함수 
def rmse_val(predict_result,y):
    return rmse(np.array(predict_result),np.array(y))


In [16]:
# 예측 값이 들어갈 빈 리스트 
predict_result=[]
#모든 x값을 한 번씩 대입하여 
for i in range(len(x)):
    #predict_result 리스트를 완성한다.
    predict_result.append(predict(x[i]))
    print("공부한 시간 =%.f, 실제 점수=%.f, 예측 점수=%.f" %(x[i],y[i],predict(x[i])))

공부한 시간 =2, 실제 점수=81, 예측 점수=82
공부한 시간 =4, 실제 점수=93, 예측 점수=88
공부한 시간 =6, 실제 점수=91, 예측 점수=94
공부한 시간 =8, 실제 점수=97, 예측 점수=100


In [17]:
#최종 RMSE출력 
print("rmse 최종값:"+str(rmse_val(predict_result,y)))

rmse 최종값:3.3166247903554


우리가 처음 가정한 a=3, b=76은 오차가 약 3.3166이라는 것을 알게되었다. 이제 남은 것은 오차를 줄이면서 새로운 선을 긋는 것이다. 
다음 장에서는 오차를 줄이는 방법인 '경사 하강법에 대해 알아보겠다.'